lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+            dRep_cmd = ' '.join(
+                [f'dRep dereplicate {dRep_workDir} --genomes {binsPooled_dir}/*.fasta'] + dRep_params)
+
+            logging.info('Running dRep command `%s`' % dRep_cmd)
+            completed_proc = subprocess.run(dRep_cmd, shell=True, stdout=sys.stdout, stderr=sys.stderr)
+
+            retcode = completed_proc.returncode
+
+            dprint('retcode', run=locals())
+
+            if retcode != 0:
+
+                # check: if Bdb.csv is empty -> nothing passed length/qual filtering
+                with open(os.path.join(dRep_workDir, 'data_tables/Bdb.csv')) as f:
+                    num_lines = sum(1 for line in f)
+
+                if num_lines == 1:
+                    msg = 'Sorry, dRep terminated with return code %d because no bins passed length and CheckM filtering' % retcode
+
+                else:
+                    msg = (
+                        "Sorry, commmand: `%s` terminated abnormally with return code: %d"
+                        % dRep_cmd, retcode
+                        )
+
+                raise Exception(msg)
+
+
+
+
+
+        #
+        ##
+        ### result BinnedContigs
+        ####
+        #####
+
+        bins_derep_dir = os.path.join(dRep_workDir, 'dereplicated_genomes')
+        objects_created = []
+
+        # for each original BinnedContigs
+        for bc in BinnedContigs.created_instances:
+            logging.info(f'Dereplicating {bc.name}')
+            bc.reduce_to_dereplicated(bins_derep_dir)
+
+            if not bc.is_empty():
+                if _globals.debug and params.get('skip_save_bc'):
+                    BinnedContigs.saved_instances.append(bc)
+                else: 
+                    objects_created.append(bc.save())
+            
+
+
+
+        #
+        ##
+        ### html
+        ####
+        #####
+
+        html_dir = os.path.join(_globals.run_dir, 'html_dir')
+        shutil.copytree('/kb/module/ui/output', html_dir) # dir of html and accessories
+
+        report.HTMLBuilder(BinnedContigs.saved_instances, dRep_cmd, dRep_workDir, html_dir)
+
+
+
+        #
+        ##
+        ### to shock
+        ####
+        #####
+
+        if _globals.debug and params.get('skip_save_shock'):
+            return
+
+
+        def dir_to_shock(dir_path, name, description):
+            '''
+            For regular directories or html directories
+            
+            name - for regular directories: the name of the flat (zip) file returned to ui
+                   for html directories: the name of the html file
+            '''
+            dfu_fileToShock_ret = _globals.dfu.file_to_shock({
+                'file_path': dir_path,
+                'make_handle': 0,
+                'pack': 'zip',
+                })
+
+            return {
+                'shock_id': dfu_fileToShock_ret['shock_id'],
+                'name': name,
+                'description': description
+                }
+
+
+        workDirZip_shockInfo = dir_to_shock(
```

Comment:
You don't need to call `file_to_shock` yourself. If you pass in the path of a directory to either html_links or file_links, then KBaseReport will zip and upload things for you. See here https://github.com/kbaseapps/KBaseReport#file-links-and-html-links